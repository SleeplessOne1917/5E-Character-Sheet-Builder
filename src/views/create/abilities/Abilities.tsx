import {
	ChangeEventHandler,
	MouseEventHandler,
	useCallback,
	useState
} from 'react';
import { ChevronDownIcon, ChevronUpIcon } from '@heroicons/react/solid';
import { addGroup, removeGroup } from '../../../redux/features/rollGroups';
import { useAppDispatch, useAppSelector } from '../../../hooks/reduxHooks';

import AbilityCalculation from '../../../components/character-creation/Abilities/AbilityCalculation';
import { AbilityItem } from '../../../types/srd';
import AbilityScores from '../../../types/abilityScores';
import ManualScores from '../../../components/character-creation/Abilities/ManualScores';
import PointBuy from '../../../components/character-creation/Abilities/PointBuy';
import RollGroup from '../../../components/character-creation/Abilities/RollGroup';
import SmallButton from '../../../components/Button/SmallButton';
import StandardArray from '../../../components/character-creation/Abilities/StandardArray';
import classes from './Abilities.module.css';
import commonClasses from '../../Views.module.css';
import { updateBase } from '../../../redux/features/abilityScores';

type AbilitiesProps = {
	abilities: AbilityItem[];
};

const Abilities = ({ abilities }: AbilitiesProps): JSX.Element => {
	const [generationMethod, setGenerationMethod] = useState('roll');
	const [showRollGroups, setShowRollGroups] = useState(true);
	const rollGroups = useAppSelector(state => state.rollGroups);
	const dispatch = useAppDispatch();

	const handleGenerationMethodChange: ChangeEventHandler<HTMLSelectElement> =
		useCallback(
			event => {
				{
					const value = event.target.value;

					if (value === 'point-buy') {
						for (const { index } of abilities) {
							dispatch(
								updateBase({ value: 8, abilityIndex: index as AbilityScores })
							);
						}
					} else {
						for (const { index } of abilities) {
							dispatch(
								updateBase({
									value: null,
									abilityIndex: index as AbilityScores
								})
							);
						}
					}

					setGenerationMethod(value);
				}
			},
			[setGenerationMethod, dispatch, abilities]
		);

	const toggleShowRollGroups: MouseEventHandler<HTMLDivElement> =
		useCallback(() => {
			setShowRollGroups(prevState => !prevState);
		}, [setShowRollGroups]);

	const addRollGroup: MouseEventHandler<HTMLButtonElement> = useCallback(() => {
		dispatch(addGroup());
	}, [dispatch]);

	const deleteRollGroup = useCallback(
		(group: number) => {
			dispatch(removeGroup({ group }));
		},
		[dispatch]
	);

	return (
		<main className={commonClasses.main}>
			<div className={commonClasses.content}>
				<h1 className={classes.title}>Ability Scores</h1>
				<div className={classes['generation-control']}>
					<label htmlFor="generation-methods">Generation Method</label>
					<select
						id="generation-methods"
						value={generationMethod}
						onChange={handleGenerationMethodChange}
					>
						<option value="manual">Manual</option>
						<option value="roll">Roll</option>
						<option value="point-buy">Point Buy</option>
						<option value="array">Standard Array</option>
					</select>
				</div>
				{generationMethod === 'manual' && (
					<ManualScores abilities={abilities} />
				)}
				{generationMethod === 'roll' && (
					<>
						<div
							className={classes['roll-groups-toggle']}
							onClick={toggleShowRollGroups}
						>
							Dice Roll Groups{' '}
							{showRollGroups ? (
								<ChevronUpIcon className={classes['roll-chevron']} />
							) : (
								<ChevronDownIcon className={classes['roll-chevron']} />
							)}
						</div>
						<div
							className={`${classes['roll-groups']}${
								showRollGroups ? ` ${classes.open}` : ''
							}`}
						>
							<div className={classes['add-group-container']}>
								<SmallButton
									positive
									style={{ marginRight: '0.5rem' }}
									onClick={addRollGroup}
								>
									+ Add Group
								</SmallButton>
								Groups: {Object.keys(rollGroups).length}
							</div>
							<div>
								{Object.keys(rollGroups).map(group => (
									<RollGroup
										abilities={abilities}
										key={group}
										onDeleteGroup={
											Object.keys(rollGroups).length > 1
												? () => deleteRollGroup(parseInt(group))
												: null
										}
										group={parseInt(group)}
									/>
								))}
							</div>
						</div>
					</>
				)}
				{generationMethod === 'point-buy' && <PointBuy abilities={abilities} />}
				{generationMethod === 'array' && (
					<StandardArray abilities={abilities} />
				)}
				<div className={classes['calculations-container']}>
					{abilities.map(ability => (
						<AbilityCalculation
							key={ability.index}
							index={ability.index}
							name={ability.full_name}
						/>
					))}
				</div>
			</div>
		</main>
	);
};

export default Abilities;
