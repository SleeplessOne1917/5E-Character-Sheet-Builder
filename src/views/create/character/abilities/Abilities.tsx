import {
	AbilityScore,
	updateBase
} from '../../../../redux/features/abilityScores';
import { ChevronDownIcon, ChevronUpIcon } from '@heroicons/react/24/solid';
import {
	GenerationMethodState,
	updateGenerationMethod
} from '../../../../redux/features/generationMethod';
import {
	KeyboardEventHandler,
	MouseEventHandler,
	useCallback,
	useState
} from 'react';
import { addGroup, removeGroup } from '../../../../redux/features/rollGroups';
import { useAppDispatch, useAppSelector } from '../../../../hooks/reduxHooks';

import AbilityCalculation from '../../../../components/character-creation/Abilities/AbilityCalculation/AbilityCalculation';
import { AbilityItem } from '../../../../types/srd';
import AbilityScores from '../../../../types/abilityScores';
import Button from '../../../../components/Button/Button';
import GeneralInfoBar from '../../../../components/character-creation/GeneralInfoBar/GeneralInfoBar';
import MainContent from '../../../../components/MainContent/MainContent';
import ManualScores from '../../../../components/character-creation/Abilities/ManualScores/ManualScores';
import PointBuy from '../../../../components/character-creation/Abilities/PointBuy/PointBuy';
import RollGroup from '../../../../components/character-creation/Abilities/Rolls/RollGroup/RollGroup';
import Select from '../../../../components/Select/Select';
import StandardArray from '../../../../components/character-creation/Abilities/StandardArray/StandardArray';
import classes from './Abilities.module.css';
import { handleKeyDownEvent } from '../../../../services/handlerService';
import { removeClassSpell } from '../../../../redux/features/classInfo';
import { removeSpell } from '../../../../redux/features/spellcasting';
import usePreparedSpells from '../../../../hooks/usePreparedSpells';

type AbilitiesProps = {
	abilities: AbilityItem[];
};

const Abilities = ({ abilities }: AbilitiesProps): JSX.Element => {
	const [showRollGroups, setShowRollGroups] = useState(true);
	const rollGroups = useAppSelector(state => state.rollGroups);
	const generationMethod = useAppSelector(state => state.generationMethod);
	const dispatch = useAppDispatch();
	const { getNumberOfSpellsToPrepare, shouldPrepareSpells } =
		usePreparedSpells();
	const classSpells = useAppSelector(
		state => state.editingCharacter.classInfo.spells
	)?.filter(({ level }) => level > 0);
	const spellcastingAbility = useAppSelector(
		state =>
			state.editingCharacter.classInfo.class?.spellcasting?.spellcasting_ability
				.index
	);
	const abilityScores = useAppSelector(
		state => state.editingCharacter.abilityScores
	);

	const removePreparedSpells = useCallback(
		(value: number | null, abilityScoreIndex: AbilityScores) => {
			if (
				spellcastingAbility &&
				abilityScoreIndex === spellcastingAbility &&
				shouldPrepareSpells
			) {
				const newPreparedSpellsNumber = getNumberOfSpellsToPrepare({
					abilityScore: {
						...(
							abilityScores as {
								[key: string]: AbilityScore;
							}
						)[abilityScoreIndex],
						base: value
					}
				});

				if (classSpells && newPreparedSpellsNumber < classSpells.length) {
					for (
						let i = 0;
						i < classSpells.length - newPreparedSpellsNumber;
						++i
					) {
						const spellToRemove = classSpells[classSpells.length - (i + 1)];

						dispatch(removeSpell(spellToRemove.index));
						dispatch(removeClassSpell(spellToRemove.index));
					}
				}
			}
		},
		[
			abilityScores,
			classSpells,
			dispatch,
			getNumberOfSpellsToPrepare,
			shouldPrepareSpells,
			spellcastingAbility
		]
	);

	const handleGenerationMethodChange = useCallback(
		(value: string) => {
			{
				if (value === 'point-buy') {
					for (const { index } of abilities) {
						dispatch(updateBase({ value: 8, abilityIndex: index }));

						removePreparedSpells(8, index);
					}
				} else {
					for (const { index } of abilities) {
						dispatch(
							updateBase({
								value: null,
								abilityIndex: index
							})
						);

						removePreparedSpells(null, index);
					}
				}

				dispatch(
					updateGenerationMethod({
						generationMethod: value as GenerationMethodState
					})
				);
			}
		},
		[dispatch, abilities, removePreparedSpells]
	);

	const toggleShowRollGroups = useCallback(() => {
		setShowRollGroups(prevState => !prevState);
	}, [setShowRollGroups]);

	const toggleShowRollGroupsKeyDown: KeyboardEventHandler<HTMLDivElement> =
		useCallback(
			event => {
				handleKeyDownEvent<HTMLDivElement>(event, toggleShowRollGroups);
			},
			[toggleShowRollGroups]
		);

	const addRollGroup: MouseEventHandler<HTMLButtonElement> = useCallback(() => {
		dispatch(addGroup());
	}, [dispatch]);

	const deleteRollGroup = useCallback(
		(group: number) => {
			dispatch(removeGroup({ group }));
		},
		[dispatch]
	);

	return (
		<MainContent testId="abilities">
			<GeneralInfoBar />
			<h1 className={classes.title}>Ability Scores</h1>
			<div className={classes['generation-control']}>
				<label id="generation-methods-label">Generation Method</label>
				<Select
					testId="generation-method"
					labelledBy="generation-methods"
					value={generationMethod}
					onChange={value => handleGenerationMethodChange(value as string)}
					options={[
						{ value: 'manual', label: 'Manual' },
						{ value: 'roll', label: 'Roll' },
						{ value: 'point-buy', label: 'Point Buy' },
						{ value: 'array', label: 'Standard Array' }
					]}
				/>
			</div>
			{generationMethod === 'manual' && <ManualScores abilities={abilities} />}
			{generationMethod === 'roll' && (
				<>
					<div
						className={classes['roll-groups-toggle']}
						onClick={toggleShowRollGroups}
						onKeyDown={toggleShowRollGroupsKeyDown}
						tabIndex={0}
						aria-label={`${showRollGroups ? 'Hide' : 'Show'} Roll Groups`}
					>
						Dice Roll Groups{' '}
						{showRollGroups ? (
							<ChevronUpIcon className={classes['roll-chevron']} />
						) : (
							<ChevronDownIcon className={classes['roll-chevron']} />
						)}
					</div>
					<div
						className={`${classes['roll-groups']}${
							showRollGroups ? ` ${classes.open}` : ''
						}`}
						data-testid="roll-groups"
					>
						<div className={classes['add-group-container']}>
							<Button positive size="small" spacing={2} onClick={addRollGroup}>
								+ Add Group
							</Button>
							Groups: {Object.keys(rollGroups).length}
						</div>
						<div>
							{Object.keys(rollGroups).map(group => (
								<RollGroup
									abilities={abilities}
									key={group}
									onDeleteGroup={
										Object.keys(rollGroups).length > 1
											? () => deleteRollGroup(parseInt(group))
											: null
									}
									group={parseInt(group)}
								/>
							))}
						</div>
					</div>
				</>
			)}
			{generationMethod === 'point-buy' && <PointBuy abilities={abilities} />}
			{generationMethod === 'array' && <StandardArray abilities={abilities} />}
			<div className={classes['calculations-container']}>
				{abilities.map(ability => (
					<AbilityCalculation
						key={ability.index}
						index={ability.index}
						name={ability.full_name}
					/>
				))}
			</div>
		</MainContent>
	);
};

export default Abilities;
