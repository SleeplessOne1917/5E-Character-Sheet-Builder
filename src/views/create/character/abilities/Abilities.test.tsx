import '@testing-library/jest-dom';

import * as stories from './Abilities.stories';

import { render, screen } from '@testing-library/react';

import Abilities from './Abilities';
import AbilityScores from '../../../../types/abilityScores';
import { Provider } from 'react-redux';
import { composeStories } from '@storybook/testing-react';
import { getAbilityScoresTest } from '../../../../hooks/useGetAbilityScore';
import { getTestStore } from '../../../../redux/store';
import { mockAbilities } from '../../../../components/character-creation/Abilities/MockAbilitiesStore';
import userEvent from '@testing-library/user-event';

const { Default } = composeStories(stories);

it('renders correctly', () => {
	render(<Default />);

	expect(screen.getByTestId('abilities')).toMatchSnapshot();
});

describe('hides and shows roll groups', () => {
	it('on click', async () => {
		render(<Default />);

		expect(screen.getByTestId('roll-groups')).toHaveClass('open');

		await userEvent.click(screen.getByText(/Dice Roll Groups/i));

		expect(screen.getByTestId('roll-groups')).not.toHaveClass('open');

		await userEvent.click(screen.getByText(/Dice Roll Groups/i));

		expect(screen.getByTestId('roll-groups')).toHaveClass('open');
	});

	it('on enter press', async () => {
		render(<Default />);

		expect(screen.getByTestId('roll-groups')).toHaveClass('open');

		await userEvent.tab();
		await userEvent.tab();
		await userEvent.keyboard('{Enter}');

		expect(screen.getByTestId('roll-groups')).not.toHaveClass('open');

		await userEvent.keyboard('{Enter}');

		expect(screen.getByTestId('roll-groups')).toHaveClass('open');
	});

	it('on space press', async () => {
		render(<Default />);

		expect(screen.getByTestId('roll-groups')).toHaveClass('open');

		await userEvent.tab();
		await userEvent.tab();
		await userEvent.keyboard(' ');

		expect(screen.getByTestId('roll-groups')).not.toHaveClass('open');

		await userEvent.keyboard(' ');

		expect(screen.getByTestId('roll-groups')).toHaveClass('open');
	});
});

it('add group adds group', async () => {
	render(<Default />);

	await userEvent.click(screen.getByText(/Add Group/i));

	expect(screen.getAllByTestId('roll-group').length).toBeGreaterThan(1);
});

it('has expected default ability score values on generation method switch', async () => {
	const store = getTestStore();
	render(
		<Provider store={store}>
			<Abilities abilities={mockAbilities} />
		</Provider>
	);

	for (const key of Object.keys(
		store.getState().editingCharacter.abilityScores
	)) {
		expect(
			getAbilityScoresTest(
				store.getState().editingCharacter.abilityScores,
				key as AbilityScores
			).base
		).toBeUndefined();
	}

	await userEvent.selectOptions(
		screen.getByTestId('generation-method-list'),
		'Manual'
	);

	for (const key of Object.keys(
		store.getState().editingCharacter.abilityScores
	)) {
		expect(
			getAbilityScoresTest(
				store.getState().editingCharacter.abilityScores,
				key as AbilityScores
			).base
		).toBeNull();
	}

	await userEvent.selectOptions(
		screen.getByTestId('generation-method-list'),
		'Roll'
	);

	for (const key of Object.keys(
		store.getState().editingCharacter.abilityScores
	)) {
		expect(
			getAbilityScoresTest(
				store.getState().editingCharacter.abilityScores,
				key as AbilityScores
			).base
		).toBeNull();
	}

	await userEvent.selectOptions(
		screen.getByTestId('generation-method-list'),
		'Point Buy'
	);

	for (const key of Object.keys(
		store.getState().editingCharacter.abilityScores
	)) {
		expect(
			getAbilityScoresTest(
				store.getState().editingCharacter.abilityScores,
				key as AbilityScores
			).base
		).toBe(8);
	}

	await userEvent.selectOptions(
		screen.getByTestId('generation-method-list'),
		'Standard Array'
	);

	for (const key of Object.keys(
		store.getState().editingCharacter.abilityScores
	)) {
		expect(
			getAbilityScoresTest(
				store.getState().editingCharacter.abilityScores,
				key as AbilityScores
			).base
		).toBeNull();
	}
});
