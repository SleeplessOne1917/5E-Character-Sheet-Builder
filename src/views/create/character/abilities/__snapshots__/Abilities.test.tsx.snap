// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<main
  class="main"
  data-testid="abilities"
>
  <div
    class="content"
  >
    <h1
      class="title"
    >
      Ability Scores
    </h1>
    <div
      class="generation-control"
    >
      <label
        id="generation-methods-label"
      >
        Generation Method
      </label>
      <div
        aria-labelledby="generation-methods"
        class="container"
        data-testid="generation-method"
      >
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          class="button"
          data-testid="generation-method-button"
          type="button"
        >
          Roll
          <svg
            aria-hidden="true"
            class="icon"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <ul
          aria-activedescendant="Roll"
          class="list"
          data-testid="generation-method-list"
          role="listbox"
          style="display: none;"
          tabindex="-1"
        >
          <li
            aria-selected="false"
            class="item"
            data-testid="generation-method-data-manual"
            role="option"
            tabindex="0"
          >
            Manual
          </li>
          <li
            aria-selected="true"
            class="item selected"
            data-testid="generation-method-data-roll"
            role="option"
            tabindex="0"
          >
            Roll
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="generation-method-data-point-buy"
            role="option"
            tabindex="0"
          >
            Point Buy
          </li>
          <li
            aria-selected="false"
            class="item"
            data-testid="generation-method-data-array"
            role="option"
            tabindex="0"
          >
            Standard Array
          </li>
        </ul>
      </div>
    </div>
    <div
      aria-label="Hide Roll Groups"
      class="roll-groups-toggle"
      tabindex="0"
    >
      Dice Roll Groups
       
      <svg
        aria-hidden="true"
        class="roll-chevron"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M14.707 12.707a1 1 0 01-1.414 0L10 9.414l-3.293 3.293a1 1 0 01-1.414-1.414l4-4a1 1 0 011.414 0l4 4a1 1 0 010 1.414z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <div
      class="roll-groups open"
      data-testid="roll-groups"
    >
      <div
        class="add-group-container"
      >
        <button
          class="button positive"
          style="font-size: 0.7rem; padding: 0.3rem; margin: 0.5rem;"
          type="button"
        >
          + Add Group
        </button>
        Groups: 
        1
      </div>
      <div>
        <div
          class="roll-group"
          data-testid="roll-group"
        >
          <div
            class="rolls"
          >
            <div
              class="display-container"
              data-testid="roll-display"
            >
              <div
                class="total"
              >
                —
              </div>
              <button
                class="button positive"
                style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
                type="button"
              >
                Roll
              </button>
            </div>
            <div
              class="display-container"
              data-testid="roll-display"
            >
              <div
                class="total"
              >
                —
              </div>
              <button
                class="button positive"
                style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
                type="button"
              >
                Roll
              </button>
            </div>
            <div
              class="display-container"
              data-testid="roll-display"
            >
              <div
                class="total"
              >
                —
              </div>
              <button
                class="button positive"
                style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
                type="button"
              >
                Roll
              </button>
            </div>
            <div
              class="display-container"
              data-testid="roll-display"
            >
              <div
                class="total"
              >
                —
              </div>
              <button
                class="button positive"
                style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
                type="button"
              >
                Roll
              </button>
            </div>
            <div
              class="display-container"
              data-testid="roll-display"
            >
              <div
                class="total"
              >
                —
              </div>
              <button
                class="button positive"
                style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
                type="button"
              >
                Roll
              </button>
            </div>
            <div
              class="display-container"
              data-testid="roll-display"
            >
              <div
                class="total"
              >
                —
              </div>
              <button
                class="button positive"
                style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem;"
                type="button"
              >
                Roll
              </button>
            </div>
          </div>
          <div>
            <button
              class="button negative"
              disabled=""
              style="font-size: 0.7rem; padding: 0.3rem; margin: 1rem;"
              type="button"
            >
              Reset Group
            </button>
            <button
              class="button positive"
              disabled=""
              style="font-size: 0.7rem; padding: 0.3rem; margin: 0.25rem;"
              type="button"
            >
              Apply Ability Scores
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="calculations-container"
    >
      <div
        class="calculation"
        data-testid="ability-calculation"
      >
        <div
          class="header"
        >
          <svg
            class="icon"
          >
            <use
              xlink:href="/Icons.svg#str"
            />
          </svg>
          Strength
        </div>
        <div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Total Score
            </div>
            <div
              class="value"
              data-testid="total-score"
            >
              —
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Modifier
            </div>
            <div
              class="value"
              data-testid="modifier"
            >
              —
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Base Score
            </div>
            <div
              class="value"
              data-testid="base-score"
            >
              —
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Racial Bonus
            </div>
            <div
              class="value"
              data-testid="racial-bonus"
            >
              +
              0
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Ability Improvements
            </div>
            <div
              class="value"
              data-testid="ability-improvements"
            >
              +
              0
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Misc Bonus
            </div>
            <div
              class="value"
              data-testid="misc-bonus"
            >
              +
              0
            </div>
          </div>
          <div
            class="component"
          >
            <div
              class="label"
            >
              <label
                for="other-str"
              >
                Other Bonus
              </label>
            </div>
            <div
              class="value input-value"
            >
              <input
                id="other-str"
                placeholder="—"
                type="number"
                value=""
              />
            </div>
          </div>
          <div
            class="component"
          >
            <div
              class="label"
            >
              <label
                for="override-str"
              >
                Override Score
              </label>
            </div>
            <div
              class="value input-value"
            >
              <input
                id="override-str"
                placeholder="—"
                type="number"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="calculation"
        data-testid="ability-calculation"
      >
        <div
          class="header"
        >
          <svg
            class="icon"
          >
            <use
              xlink:href="/Icons.svg#dex"
            />
          </svg>
          Dexterity
        </div>
        <div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Total Score
            </div>
            <div
              class="value"
              data-testid="total-score"
            >
              —
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Modifier
            </div>
            <div
              class="value"
              data-testid="modifier"
            >
              —
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Base Score
            </div>
            <div
              class="value"
              data-testid="base-score"
            >
              —
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Racial Bonus
            </div>
            <div
              class="value"
              data-testid="racial-bonus"
            >
              +
              0
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Ability Improvements
            </div>
            <div
              class="value"
              data-testid="ability-improvements"
            >
              +
              0
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Misc Bonus
            </div>
            <div
              class="value"
              data-testid="misc-bonus"
            >
              +
              0
            </div>
          </div>
          <div
            class="component"
          >
            <div
              class="label"
            >
              <label
                for="other-dex"
              >
                Other Bonus
              </label>
            </div>
            <div
              class="value input-value"
            >
              <input
                id="other-dex"
                placeholder="—"
                type="number"
                value=""
              />
            </div>
          </div>
          <div
            class="component"
          >
            <div
              class="label"
            >
              <label
                for="override-dex"
              >
                Override Score
              </label>
            </div>
            <div
              class="value input-value"
            >
              <input
                id="override-dex"
                placeholder="—"
                type="number"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="calculation"
        data-testid="ability-calculation"
      >
        <div
          class="header"
        >
          <svg
            class="icon"
          >
            <use
              xlink:href="/Icons.svg#con"
            />
          </svg>
          Constitution
        </div>
        <div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Total Score
            </div>
            <div
              class="value"
              data-testid="total-score"
            >
              —
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Modifier
            </div>
            <div
              class="value"
              data-testid="modifier"
            >
              —
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Base Score
            </div>
            <div
              class="value"
              data-testid="base-score"
            >
              —
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Racial Bonus
            </div>
            <div
              class="value"
              data-testid="racial-bonus"
            >
              +
              0
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Ability Improvements
            </div>
            <div
              class="value"
              data-testid="ability-improvements"
            >
              +
              0
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Misc Bonus
            </div>
            <div
              class="value"
              data-testid="misc-bonus"
            >
              +
              0
            </div>
          </div>
          <div
            class="component"
          >
            <div
              class="label"
            >
              <label
                for="other-con"
              >
                Other Bonus
              </label>
            </div>
            <div
              class="value input-value"
            >
              <input
                id="other-con"
                placeholder="—"
                type="number"
                value=""
              />
            </div>
          </div>
          <div
            class="component"
          >
            <div
              class="label"
            >
              <label
                for="override-con"
              >
                Override Score
              </label>
            </div>
            <div
              class="value input-value"
            >
              <input
                id="override-con"
                placeholder="—"
                type="number"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="calculation"
        data-testid="ability-calculation"
      >
        <div
          class="header"
        >
          <svg
            class="icon"
          >
            <use
              xlink:href="/Icons.svg#int"
            />
          </svg>
          Intelligence
        </div>
        <div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Total Score
            </div>
            <div
              class="value"
              data-testid="total-score"
            >
              —
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Modifier
            </div>
            <div
              class="value"
              data-testid="modifier"
            >
              —
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Base Score
            </div>
            <div
              class="value"
              data-testid="base-score"
            >
              —
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Racial Bonus
            </div>
            <div
              class="value"
              data-testid="racial-bonus"
            >
              +
              0
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Ability Improvements
            </div>
            <div
              class="value"
              data-testid="ability-improvements"
            >
              +
              0
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Misc Bonus
            </div>
            <div
              class="value"
              data-testid="misc-bonus"
            >
              +
              0
            </div>
          </div>
          <div
            class="component"
          >
            <div
              class="label"
            >
              <label
                for="other-int"
              >
                Other Bonus
              </label>
            </div>
            <div
              class="value input-value"
            >
              <input
                id="other-int"
                placeholder="—"
                type="number"
                value=""
              />
            </div>
          </div>
          <div
            class="component"
          >
            <div
              class="label"
            >
              <label
                for="override-int"
              >
                Override Score
              </label>
            </div>
            <div
              class="value input-value"
            >
              <input
                id="override-int"
                placeholder="—"
                type="number"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="calculation"
        data-testid="ability-calculation"
      >
        <div
          class="header"
        >
          <svg
            class="icon"
          >
            <use
              xlink:href="/Icons.svg#wis"
            />
          </svg>
          Wisdom
        </div>
        <div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Total Score
            </div>
            <div
              class="value"
              data-testid="total-score"
            >
              —
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Modifier
            </div>
            <div
              class="value"
              data-testid="modifier"
            >
              —
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Base Score
            </div>
            <div
              class="value"
              data-testid="base-score"
            >
              —
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Racial Bonus
            </div>
            <div
              class="value"
              data-testid="racial-bonus"
            >
              +
              0
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Ability Improvements
            </div>
            <div
              class="value"
              data-testid="ability-improvements"
            >
              +
              0
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Misc Bonus
            </div>
            <div
              class="value"
              data-testid="misc-bonus"
            >
              +
              0
            </div>
          </div>
          <div
            class="component"
          >
            <div
              class="label"
            >
              <label
                for="other-wis"
              >
                Other Bonus
              </label>
            </div>
            <div
              class="value input-value"
            >
              <input
                id="other-wis"
                placeholder="—"
                type="number"
                value=""
              />
            </div>
          </div>
          <div
            class="component"
          >
            <div
              class="label"
            >
              <label
                for="override-wis"
              >
                Override Score
              </label>
            </div>
            <div
              class="value input-value"
            >
              <input
                id="override-wis"
                placeholder="—"
                type="number"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="calculation"
        data-testid="ability-calculation"
      >
        <div
          class="header"
        >
          <svg
            class="icon"
          >
            <use
              xlink:href="/Icons.svg#cha"
            />
          </svg>
          Charisma
        </div>
        <div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Total Score
            </div>
            <div
              class="value"
              data-testid="total-score"
            >
              —
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Modifier
            </div>
            <div
              class="value"
              data-testid="modifier"
            >
              —
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Base Score
            </div>
            <div
              class="value"
              data-testid="base-score"
            >
              —
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Racial Bonus
            </div>
            <div
              class="value"
              data-testid="racial-bonus"
            >
              +
              0
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Ability Improvements
            </div>
            <div
              class="value"
              data-testid="ability-improvements"
            >
              +
              0
            </div>
          </div>
          <div
            class="component calculation-component"
          >
            <div
              class="label calculation-label"
            >
              Misc Bonus
            </div>
            <div
              class="value"
              data-testid="misc-bonus"
            >
              +
              0
            </div>
          </div>
          <div
            class="component"
          >
            <div
              class="label"
            >
              <label
                for="other-cha"
              >
                Other Bonus
              </label>
            </div>
            <div
              class="value input-value"
            >
              <input
                id="other-cha"
                placeholder="—"
                type="number"
                value=""
              />
            </div>
          </div>
          <div
            class="component"
          >
            <div
              class="label"
            >
              <label
                for="override-cha"
              >
                Override Score
              </label>
            </div>
            <div
              class="value input-value"
            >
              <input
                id="override-cha"
                placeholder="—"
                type="number"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</main>
`;
